File: java/formula-java/src/main/java/org/formula/FormulaRandom.java
package org.formula;

// const M = 0x80000000;
//const A = 1103515245;
//const C = 12345;

public class FormulaRandom {
    private static final long M = 2147483647L;
    private static final long A = 1103515245L;
    private static final long C = 12345L;
    private static final long MAX_SEED = Math.floorDiv(0xFFFFL, A) - C - 1;
    private long state;

    public FormulaRandom() {
        this((long) Math.floor(Math.random() * (M - 1)));
    }

    public FormulaRandom(long seed) {
        this.state = seed % MAX_SEED;
    }

    public long nextInt() {
        var result = (A * this.state + C) % M;
        this.state = result % MAX_SEED;
        return result;
    }

    public float nextFloat() {
        return this.nextInt() / (float)(M - 1);
    }
}

File: js/formula-js/src/FormulaRandom.ts
const M = 2147483647;
const A = 1103515245;
const C = 12345;

const MAX_SEED = Math.floor(0xFFFF / A) - C - 1;

export class FormulaRandom {
  private state: number;

  constructor(seed?: number) {
    this.state = (seed !== undefined
      ? seed
      : Math.floor(Math.random() * (M - 1))) % MAX_SEED;
  }

  public nextInt(): number {
    const result = (A * this.state + C) % M;
    this.state = result % MAX_SEED;
    return result;
  }

  public nextFloat(): number {
    return this.nextInt() / (M - 1);
  }

  public nextRange(start: number, end: number): number {
    var rangeSize = end - start;
    var randomUnder1 = this.nextInt() / M;
    return start + Math.floor(randomUnder1 * rangeSize);
  }
}